### PR TITLE
Union rest mapper

### PR DESCRIPTION
**Pull request by deads2k**
Includes https://github.com/kubernetes/kubernetes/pull/20814.

This allows a proper combination of `RESTMappers` so that all responses are proper unions of all constituent `RESTMappers` instead of "first hit wins".  This makes combination `RESTMappers` easier to put together.

If we still want FirstOneWins RESTMappers, I'm willing to refactor this into a separate struct.  I can see valid cases for the other way, but this is what we've wanted by default downstream.

### CODE REVIEW

**Comment by deads2k** [comment]
@caesarxuchao @smarterclayton I'll bet both of you have opinions.

**Comment by bgrant0607** [comment]
cc @kubernetes/kubectl @kubernetes/goog-csi

**Comment by kargakis** [comment]
LGTM

File: pkg/api/meta/multirestmapper.go [hunk 1]
```diff
@@ -50,72 +51,148 @@ func (m MultiRESTMapper) ResourceSingularizer(resource string) (singular string,
...
+func (m MultiRESTMapper) RESTMapping(gk unversioned.GroupKind, versions ...string) (*RESTMapping, error) {
+	allMappings := []*RESTMapping{}
+	errors := []error{}
+
 	for _, t := range m {
-		mapping, err = t.RESTMapping(gk, versions...)
-		if err == nil {
-			return
+		currMapping, err := t.RESTMapping(gk, versions...)
+		// ignore "no match" errors, but any other error percolates back up
```
Comment (smarterclayton) [inline pkg/api/meta/multirestmapper.go]: Comment is wrong - did you intend to handle no match?

File: pkg/kubectl/kubectl.go [hunk 2]
```diff
@@ -72,51 +71,71 @@ func (m OutputVersionMapper) RESTMapping(gk unversioned.GroupKind, versions ...s
...
+func (e ShortcutExpander) RESTMapping(gk unversioned.GroupKind, versions ...string) (*meta.RESTMapping, error) {
+	return e.RESTMapper.RESTMapping(gk, versions...)
+}
+
+func (e ShortcutExpander) AliasesForResource(resource string) ([]string, bool) {
+	return e.RESTMapper.AliasesForResource(expandResourceShortcut(unversioned.GroupVersionResource{Resource: resource}).Resource)
+}
+
+// shortForms is the list of short names to their expanded names
+var shortForms = map[string]string{
```
Comment (smarterclayton) [inline pkg/kubectl/kubectl.go]: Did you move this in the other PR?

File: pkg/kubectl/resource/builder.go [hunk 3]
```diff
@@ -440,15 +440,27 @@ func (b *Builder) resourceMappings() ([]*meta.RESTMapping, error) {
...
-		gvk, err := b.mapper.KindFor(unversioned.GroupVersionResource{Resource: r})
+		gvks, err := b.mapper.KindsFor(unversioned.GroupVersionResource{Resource: r})
 		if err != nil {
 			return nil, err
 		}
-		mapping, err := b.mapper.RESTMapping(gvk.GroupKind(), gvk.Version)
-		if err != nil {
-			return nil, err
+		// the list is in most-preferred to least preferred, so iterate until we find a hit
+		var firstErr error
```
Comment (smarterclayton) [inline pkg/kubectl/resource/builder.go]: Probably this should be abstracted into a method `getFirstMapping()` and reused here and below.

File: pkg/api/meta/multirestmapper.go (hunk 1 above)
Comment (deads2k) [inline pkg/api/meta/multirestmapper.go]: > Comment is wrong - did you intend to handle no match?

Yes, it makes for a cleaner error message.  I will update.

File: pkg/kubectl/resource/builder.go (hunk 3 above)
Comment (deads2k) [inline pkg/kubectl/resource/builder.go]: > Probably this should be abstracted into a method getFirstMapping() and reused here and below.

When this pull gets combined with PriorityRESTMapper and the disambiguation pull, I'll be able to remove all changes to `resource/builder` and I expect we'll be merging them all at once.

Comment (lavalamp) [inline pkg/kubectl/resource/builder.go]: +1

**Comment by lavalamp** [comment]
I left comments on the commits. Minor stuff, then lgtm

**Comment by deads2k** [comment]
> I left comments on the commits. Minor stuff, then lgtm

Ok, I'll update for comments and combine into a single pull for final review.

**Comment by deads2k** [comment]
superceded by https://github.com/kubernetes/kubernetes/pull/21272

**Pull request closed without merging**